ДЗ "Homework1 - Coroutines": Виталий Макаркин

VitalyMakarkin opened the pull request.

Commit:
Move from callable to suspend

Commit:
Add image service

Commit:
Change to ViewModel, add CoroutineExceptionHandler, add Result class

antonkazakov commented on app/src/main/java/otus/homework/coroutines/PresenterScope.kt:
```diff
@@ -0,0 +1,12 @@
+class PresenterScope : CoroutineScope {
+
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.Main + CoroutineName("CatsCoroutine")
```
Добавь здесь инстанс `Job`/`SupervisorJob`

VitalyMakarkin replied:
Добавил

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,35 +1,51 @@
     }
 
     fun detachView() {
+        scope.cancel()
```
Вроде здесь будет краш, потому что нет инстанса Job в самом скоупе

VitalyMakarkin replied:
Поправил выше, теперь ок

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,35 +1,51 @@
+                )
+
+                _catsView?.populate(factWithImage)
+            } catch (e: Exception) {
```
Точно у тебя здесь `catch` блок сработает? Проверь, пробрось исключение в имплементации `catsService`

VitalyMakarkin replied:
Пробросил

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsPresenter.kt:
```diff
@@ -1,35 +1,51 @@
+                        _catsView?.showShortToast(R.string.server_not_responding)
+                    }
+
+                    else -> {
```
`CancellationException` неообходимо дальше пробросить, либо catch сделать только на `SocketTimeoutException`

VitalyMakarkin replied:
Не стал трогать presenter после введения viewModel, но поправил

antonkazakov commented on app/src/main/java/otus/homework/coroutines/CatsViewModel.kt:
```diff
@@ -0,0 +1,68 @@
+            CrashMonitor.trackWarning()
+        }) {
+            try {
+                val factWithImage = withContext(Dispatchers.IO) {
```
Можно не переключать контекст. Ретрофит из коробки поддерживает саспенд функции и не блокирует поток

VitalyMakarkin replied:
Согласен, убрал, все работает

antonkazakov reviewed (changes requested):
Привет. пару замечаний оставил

antonkazakov reviewed (changes requested):
Привет. пару замечаний оставил

Commit:
Fix
- add SupervisorJob for custom PresenterScope
- remove withContext
- refactor exception handling for presenter

antonkazakov reviewed (commented):
👍
